File: Codechef/NOV-19B/WEIRDO.py
import sys

v = {"a", "e", "i", "o", "u"}

def power(b, p):
    res = 1

    while p:
        if p & 0x1: res *= b
        b *= b
        p >>= 1

    return res

def createdBy(s):
  ln = len(s)
  for i in range(ln - 2):
    cnt = int(s[i] in v) + int(s[i + 1] in v) + int(s[i + 2] in v)
    if cnt < 2:
      return False
  return True

def main():
  sys.stdin = open('in.txt', 'r')
  # sys.stdout = open('out.txt', 'w')
  t = int(sys.stdin.readline().strip())
  for _ in range(t):
    n, m = 0, 0
    ax, afx, bx, bfx = {}, {}, {}, {}
    ln = int(sys.stdin.readline().strip())

    for _ in range(ln):
      s = sys.stdin.readline().strip()
      alpha = set()
      alpha.clear()

      if (createdBy(s)):
        n += 1
        for ch in s:
          afx.setdefault(ch, 0)
          afx[ch] = afx[ch] + 1
          if ch not in alpha:
            ax.setdefault(ch, 0)
            ax[ch] += 1
            alpha.add(ch)
      else:
        m += 1
        for ch in s:
          bfx.setdefault(ch, 0)
          bfx[ch] += 1
          if ch not in alpha:
            bx.setdefault(ch, 0)
            bx[ch] += 1
            alpha.add(ch)

    sca, scb = 1.0, 1.0
    
    for ch in ax:
      num = power(afx[ch], n)
      sca = sca * (ax[ch] / num)
    
    for ch in bx:
      num = power(bfx[ch], m)
      scb = scb * (bx[ch] / num)

    ans = sca / scb
    
    print("Infinity") if (ans > 1e7) else print(ans)
    

if __name__ == "__main__":
  main()

File: Codechef/NOV-19B/WEIRDO.cpp
#pragma GCC target("avx")
#pragma GCC optimization("O3")
#pragma GCC optimization("unroll-loops")

#include <bits/stdc++.h>
#include <ext/pb_ds/assoc_container.hpp>
using namespace __gnu_pbds;
using namespace std;

typedef long long ll;
#define ff first
#define ss second
#define pii pair<int, int>
#define pll pair<ll, ll>
#define sz(a) a.size()
#define all(a) a.begin(), a.end()
#define w(a) cerr << #a << " : " << (a) << "\n";

unordered_set<char> v({'a', 'e', 'i', 'o', 'u'});

double pow(ll base, ll exp) {
  double result = 1.0;
  for (;;) {
    if (exp & 1) result *= base;
    exp >>= 1;
    if (!exp) break;
    base *= base;
  }
  return result;
}

bool createdBy(string &s) {
  ll n = sz(s);
  if (n == 2) {
    int cnt = v.count(s[0]) + v.count(s[1]);
    if (cnt == 0) return false;
  } else {
    for (register int i = 0; i <= n - 3; i++) {
      int cnt = v.count(s[i]) + v.count(s[i + 1]) + v.count(s[i + 2]);
      if (cnt < 2) return false;
    }
  }
  return true;
}

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(0);
  cout.precision(7);
#ifndef ONLINE_JUDGE
  freopen("in.txt", "r", stdin);
  // freopen("out.txt", "w", stdout);
#endif
  int t;
  cin >> t;
  while (t--) {
    ll sn;
    cin >> sn;
    gp_hash_table<char, double> ax, afx, bx, bfx;
    double n = 0.0, m = 0.0;

    for (register int i = 0; i < sn; i++) {
      string s;
      cin >> s;

      if (createdBy(s)) {  // Alice
        bitset<32> b;
        n++;
        for (char ch : s) {
          afx[ch]++;
          if (b[ch - 96] == 0) {
            ax[ch]++;
            b[ch - 96] = 1;
          }
        }
      } else {  // Bob
        bitset<27> b;
        m++;
        for (char ch : s) {
          bfx[ch]++;
          if (b[ch - 96] == 0) {
            bx[ch]++;
            b[ch - 96] = 1;
          }
        }
      }
    }

    double sca = 1.0, scb = 1.0;
    for (auto d : ax) sca = sca * (d.ss / pow(afx[d.ff], n));
    for (auto d : bx) scb = scb * (d.ss / pow(bfx[d.ff], m));
    double ans = sca / scb;
    (ans > 1e7) ? cout << "Infinity"
                       << "\n"
                : cout << fixed << ans << "\n";
  }
  return 0;
}
